xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java

```diff
@@ -0,0 +1,140 @@
+import io.grpc.ClientStreamTracer;
+import io.grpc.ClientStreamTracer.StreamInfo;
+import io.grpc.LoadBalancer.PickResult;
+import io.grpc.Metadata;
+import io.grpc.xds.XdsClientLoadRecorder.ClientLoadCounter;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An {@link XdsLoadReportStore} instance holds the client side load stats for a cluster.
+ */
+@ThreadSafe
+final class XdsLoadReportStore {
+
+  private final String clusterName;
+  private final ConcurrentMap<Locality, ClientLoadCounter> localityLoadCounters;
+  // Cluster level dropped request counts for each category specified in the DropOverload policy.
+  private final ConcurrentMap<String, AtomicLong> dropCounters;
+
+  XdsLoadReportStore(String clusterName) {
+    this(clusterName, new ConcurrentHashMap<Locality, ClientLoadCounter>(),
+        new ConcurrentHashMap<String, AtomicLong>());
+  }
+
+  @VisibleForTesting
+  XdsLoadReportStore(String clusterName,
+      ConcurrentMap<Locality, ClientLoadCounter> localityLoadCounters,
+      ConcurrentMap<String, AtomicLong> dropCounters) {
+    this.clusterName = checkNotNull(clusterName, "clusterName");
+    this.localityLoadCounters = checkNotNull(localityLoadCounters, "localityLoadCounters");
+    this.dropCounters = checkNotNull(dropCounters, "dropCounters");
+  }
+
+  /**
+   * Generates a {@link ClusterStats} containing load stats in locality granularity.
+   */
+  ClusterStats generateLoadReport(Duration interval) {
+    ClusterStats.Builder statsBuilder = ClusterStats.newBuilder().setClusterName(clusterName)
+        .setLoadReportInterval(interval);
+    for (Map.Entry<Locality, XdsClientLoadRecorder.ClientLoadCounter> entry : localityLoadCounters
+        .entrySet()) {
+      XdsClientLoadRecorder.ClientLoadSnapshot snapshot = entry.getValue().snapshot();
+      statsBuilder
+          .addUpstreamLocalityStats(UpstreamLocalityStats.newBuilder()
+              .setLocality(entry.getKey())
+              .setTotalSuccessfulRequests(snapshot.callsSucceed)
+              .setTotalErrorRequests(snapshot.callsFailed)
+              .setTotalRequestsInProgress(snapshot.callsInProgress));
+    }
+    for (Map.Entry<String, AtomicLong> entry : dropCounters.entrySet()) {
+      statsBuilder.addDroppedRequests(DroppedRequests.newBuilder()
+          .setCategory(entry.getKey())
+          .setDroppedCount(entry.getValue().get()));
+    }
+    return statsBuilder.build();
+  }
+
+  /**
+   * Create a {@link ClientLoadCounter} for the provided locality if not present in this {@link
+   * XdsLoadReportStore}. This method needs to be called at locality updates for newly assigned
+   * localities in balancer discovery responses.
+   */
+  void addLocality(Locality locality) {
+    localityLoadCounters.putIfAbsent(locality, new ClientLoadCounter());
```

Comment:
The xDS balancer should be able to correctly track which localities are new. It should be considered a bug if the xDS policy call this with a locality that's already here. We should call `put()` and assert that the return value is `null` to expose such bug.

Reply:
ACK.